openfasoc/MLoptimization/run_training.py

```diff
@@ -0,0 +1,286 @@
+from gymnasium import spaces
+from gymnasium.spaces import Discrete
+from gymnasium.wrappers import EnvCompatibility
+from ray.rllib.env.wrappers.multi_agent_env_compatibility import MultiAgentEnvCompatibility
+from sky130_nist_tapeout import single_build_and_simulation
+sky130_nist_tapeout.path.append('../generators/gdsfactory-gen/')
+import numpy as np
+import random
+import psutil
+
+from multiprocessing import Pool
+from collections import OrderedDict
+import yaml
+import yaml.constructor
+import statistics
+import os
+import itertools
+import pickle
+import yaml
+from pathlib import Path
+
+#
+#environment set up
+class Envir(gym.Env):
+    metadata = {'render.modes': ['human']}
+
+    PERF_LOW = -1
+    PERF_HIGH = 1
+
+    def __init__(self, env_config):
+        PERF_LOW = -1
+        PERF_HIGH = 1
+        self.multi_goal = env_config.get("multi_goal",False)
+        self.generalize = env_config.get("generalize",False)
+        num_valid = env_config.get("num_valid",50)
+        self.specs_save = env_config.get("save_specs", False)
+        self.valid = env_config.get("run_valid", False)
+        self.horizon = env_config.get("horizon", 100)
+        inputspec = env_config.get("inputspec",{})
+
+        self.env_steps = 0
+        #data = np.load('./training_params.npy')
+        #result = np.load('./training_results.npy')
+        #self.result = result
+        self.epi_steps = 0
+
+        specs = {}
+        print("num valid isssssssssssssssssssssss:" + str(self.horizon))
+        if(self.valid):
+            print("hahahahahahaahahahahahahahahahaahah")
+        # design specs
+        if self.generalize == True:
+            if self.valid == False:
+                specs = yaml.safe_load(Path('new_spec_1.yaml').read_text())
+            else:
+                specs = inputspec
+                print(inputspec)
+                print("haaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
+
+
+        self.specs = specs
+
+        self.specs_ideal = []
+        self.specs_id = list(self.specs.keys())
+        self.fixed_goal_idx = -1
+        self.num_os = len(list(self.specs.values())[0])
+
+        # [10, 0, 0, 0, 0, 0, 0, 0, 10, 0, 5, 0, 10, 0, 0, 0, 0, 10]
+        #[10, 4, 10,  0, 0, 0, 10, 0, 10,  0, 10, 0,  0, 0, 0, 0, 0, 0]
+        #[9.  2.  6.  6. 2. 4. 9.  1. 2.   3. 10. 0.3 6. 3. 12. 12. 2. 1.]
+        #[9.  1.  6.  6. 2. 4. 9.  1. 6.   3. 10. 0.3 6. 3. 12. 12. 2. 1. ]
+
+        # param array
+        params = {
+                  "diffpair_params0" : [3, 9.6, 0.6],
+                  "diffpair_params1" : [0.3, 2.17, 0.17],
+                  "diffpair_params2" : [2, 7, 1],
+                  "houtput_bias0" : [3, 9.6, 0.6],
+                  "houtput_bias2" : [2, 7, 1],
+                  "pamp_hparams0" : [4, 10.6, 0.6],
+                  "pamp_hparams1" : [0.3, 2.1, 0.1],
+                  "pamp_hparams2" : [6, 15, 1],
+                  "mim_cap_rows" : [2, 4, 1],
+                  "rmult" : [1, 3, 1],
+                  }
```

Comment:
If too cumbersome to place as argparse commands, should at least be constants with meaningful names, e.g. MIM_CAP_ROW_MIN, MIM_CAP_ROW_MAX, MIM_CAP_ROW_STEP.